### PR TITLE
dynamic params and facets

### PR DESCRIPTION
**Pull request by larsbuntemeyer**
This is just adding my first experience with ESGF CORDEX API. It's basically just adapting request parameters and facets for different ESGF projects.

### CODE REVIEW

File: pangeo_forge_esgf/recipe_inputs.py [hunk 1]
```diff
@@ -140,6 +146,9 @@ async def _esgf_api_request(
     if params["latest"] == "true":
         if "version" in facets:
             del facets["version"]
```
Comment (jbusecke) [inline pangeo_forge_esgf/recipe_inputs.py:148]: I have previously wondered that this seems yanky. If I set "latest" to false in the requests I can actually provide the version. Is that behavior the same for CORDEX?

**Review by jbusecke — CHANGES_REQUESTED** [review]
Dear @larsbuntemeyer thank you so much for the engagement here! I really appreciate your help with this.
Your Pr has exposed some issues in my original implementation(especially the lack of tests) and made me think of more general cases (which is very great). 
I think that I would like to understand how we can generalize this code to even more mips in a general way (hopefully increasing the usefulness).
I left some detailed comments, but except for some smaller issues (hardcoding table_id=Amo) I think this Pr is a very good addition. Lets iterate on the details and then we can merge this?

File: pangeo_forge_esgf/utils.py [hunk 2]
```diff
@@ -12,3 +14,24 @@ def facets_from_iid(iid: str, mip: str = None) -> Dict[str, str]:
...
+    }
+    r = requests.get(url, params)
+    #print(r.status_code)
+    return r.json()["response"]["docs"][0]["dataset_id_template_"][0]
+
+
+def facets_from_template(template: str):
+    """Parse the (dataset_id) string template into a list of (facet) keys"""
+    regex = r"\((.*?)\)"
+    return re.findall(regex, template)
```
Comment (larsbuntemeyer) [inline pangeo_forge_esgf/utils.py:37]: I dived a little bit into the [ESGF API documentation](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html) and added now these two functions that should be able to retrieve the `dataset_id_template_` (and parse it into facets) for any ESGF project. I don't use them yet but i think this is better than what i did in the `params` module.

File: pangeo_forge_esgf/utils.py [hunk 3]
```diff
@@ -12,3 +14,24 @@ def facets_from_iid(iid: str, mip: str = None) -> Dict[str, str]:
...
+        url = "https://esgf-node.llnl.gov/esg-search/search"
+    params = {
+        "project": project,
+        "fields": "*",
+        "limit": 1,
+        "format": "application/solr+json"
+    }
+    r = requests.get(url, params)
+    #print(r.status_code)
+    return r.json()["response"]["docs"][0]["dataset_id_template_"][0]
```
Comment (larsbuntemeyer) [inline pangeo_forge_esgf/utils.py:31]: I dived a little bit into the [ESGF API documentation](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html) and added now these two functions that should be able to retrieve the `dataset_id_template_` (and parse it into facets) for any ESGF project. I don't use them yet but i think this is better than what i did in the `params` module, e.g.,
```python
template = utils.get_dataset_id_template("CMIP6")
# returns '%(mip_era)s.%(activity_drs)s.%(institution_id)s.%(source_id)s.%(experiment_id)s.%(member_id)s.%(table_id)s.%(variable_id)s.%(grid_label)s'
```

File: pangeo_forge_esgf/params.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# dataset id templates
+cmip6_template = "mip_era.activity_id.institution_id.source_id.experiment_id.variant_label.table_id.variable_id.grid_label.version"
+cordex_template = "project.product.domain.institute.driving_model.experiment.ensemble.rcm_name.rcm_version.time_frequency.variable"
```
[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/params.py:3]: I dived a little more into the [ESGF API documentation](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html) and added two functions to retrieve that information from the API now.

**Comment by larsbuntemeyer** [comment]
Thanks for your all your comment @jbusecke ! I'll clean this up now and let you know when i am ready to merge!

File: pangeo_forge_esgf/utils.py [hunk 5]
```diff
@@ -1,9 +1,37 @@
...
+
+
+def facets_from_template(template: str):
+    """Parse the (dataset_id) string template into a list of (facet) keys"""
+    regex = r"\((.*?)\)"
+    return re.findall(regex, template)
```
Comment (larsbuntemeyer) [inline pangeo_forge_esgf/utils.py:37]: I dived a little bit into the [ESGF API documentation](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html) and added now these two functions that should be able to retrieve the `dataset_id_template_` (and parse it into facets) for any ESGF project. I don't use them yet but i think this is better than what i did in the `params` module, e.g.,
```python
template = utils.get_dataset_id_template("CMIP6")
# returns '%(mip_era)s.%(activity_drs)s.%(institution_id)s.%(source_id)s.%(experiment_id)s.%(member_id)s.%(table_id)s.%(variable_id)s.%(grid_label)s'
```
which can be parsed into:
```python
utils.facets_from_template(template)
```
```
['mip_era',
 'activity_drs',
 'institution_id',
 'source_id',
 'experiment_id',
 'member_id',
 'table_id',
 'variable_id',
 'grid_label']
```

Comment (larsbuntemeyer) [reply pangeo_forge_esgf/utils.py:37]: hmmm, here we have `member_id` instead of `variant_label`, that seems not right...

File: pangeo_forge_esgf/dynamic_kwargs.py [hunk 6]
```diff
@@ -96,9 +97,11 @@ async def response_data_processing(
     session: aiohttp.ClientSession,
     response_data: List[Dict[str, str]],
     iid: str,
+    ssl: ssl.SSLContext = None,
 ) -> Tuple[List[str], Dict[str, Dict[str, str]]]:
 
-    table_id = facets_from_iid(iid).get("table_id")
+    # really hacky!
+    table_id = "Amon"  # facets_from_iid(iid).get("table_id")
```
[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/dynamic_kwargs.py:104]: Sorry, this was just a quick hack (i didn't actually want to commit that)... I think the highest level term is called `project` as done, e.g., in the [facet search example](https://esgf.github.io/esg-search/ESGF_Search_RESTful_API.html#facet-listings)...

File: pangeo_forge_esgf/params.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/params.py:3]: Yes, the time frequency in CORDEX is described in "time_frequency", the values are not 100% percent compatible but they contain basically the same vocabulary as in CMIP5, e.g., "day", "mon", "1hr", etc... However, this is no indication of the time cell method, e.g., "1hrPt" as in CMIP6...

File: test_cordex.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+sslcontext.load_cert_chain(manager.esgf_credentials)
+
+iids = [
+    "CORDEX.output.EUR-11.GERICS.ECMWF-ERAINT.evaluation.r1i1p1.REMO2015.v1.mon.tas.v20180813",
+    "CORDEX.output.EUR-44.MPI-CSC.MPI-M-MPI-ESM-LR.historical.r1i1p1.REMO2009.v1.mon.tas.v20150609"
+]
+
+recipe_inputs = asyncio.run(generate_recipe_inputs_from_iids(iids, ssl=sslcontext))
+print("DONE")
+print(recipe_inputs)
```
[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply test_cordex.py:25]: Just to mention here, that the actual recipe creation won't work with the latest version of `pange-forge-recipes` because of https://github.com/pangeo-forge/pangeo-forge-recipes/issues/418...  when having an SSLObject in the FilePattern...

File: pangeo_forge_esgf/utils.py (hunk 5 above)
Comment (jbusecke) [reply pangeo_forge_esgf/utils.py:37]: Oh this was a long standing issue of confusion for me. I believe it is only relevant for DAMIP, but I will need to double check this.

**Comment by jbusecke** [comment]
This looks really cool. Ultimately I wonder if we should substantially refactor this logic and make it more object oriented. 
Just thinking out loud: 
Should we have something like this:
```python
class ESGFBuilder:
    def __init__(project, ):
        ...defines class properties like the facets from project type via the request you outlined above...
    
    ... set up the connection to the server for the duration of the session...
    ...logic that retrieves urls etc...
    ...method that returns recipes...
```
I will have to invest some more time into this refactor, and wrapping my mind around async, but I just wanted to brain dump this here in case you had some input. 

I really like the direction this is going for, generalizing this tool for different ESGF projects!

File: README.md [hunk 8]
```diff
@@ -35,4 +35,37 @@ and you will get:
...
+You can also use different ESGF projects and facets to search, e.g., for a search in the ESGF CORDEX datasets, use:
+```python
+parse_iids = [
+    "CORDEX.output.EUR-11.MPI-CSC.MPI-M-MPI-ESM-LR.*.r1i1p1.REMO2009.v1.mon.tas",
+    "CORDEX.output.EUR-44.MPI-CSC.MPI-M-MPI-ESM-LR.*.r1i1p1.REMO2009.v1.mon.tas"
+]
+iids = []
+```
+
+using a different url:
```
Comment (jbusecke) [inline README.md:52]: Can you clarify if this is necessary or if this is just to show that the functionality works on all the ESGF search nodes?

File: pangeo_forge_esgf/dynamic_kwargs.py [hunk 9]
```diff
@@ -96,9 +97,11 @@ async def response_data_processing(
     session: aiohttp.ClientSession,
     response_data: List[Dict[str, str]],
     iid: str,
+    ssl: ssl.SSLContext = None,
```
Comment (jbusecke) [inline pangeo_forge_esgf/dynamic_kwargs.py:100]: For now this is great, but adding all these input arguments is a bit of a code smell to me. I think we should aim to set up a 'Connection' or 'Client' object once and then pass this around the different logic that needs to make a request?

File: pangeo_forge_esgf/dynamic_kwargs.py [hunk 10]
```diff
@@ -110,7 +113,7 @@ async def response_data_processing(
     # print(urls)
     print(f"{iid}: Check for netcdf 3 files")
     pattern_kwargs = {}
-    netcdf3_check = await is_netcdf3(session, urls[-1])
+    netcdf3_check = await is_netcdf3(session, urls[-1], ssl)
```
Comment (jbusecke) [inline pangeo_forge_esgf/dynamic_kwargs.py:116]: Same here.

File: pangeo_forge_esgf/recipe_inputs.py [hunk 11]
```diff
@@ -225,24 +234,25 @@ async def pick_data_node(
...
+            status = await check_url(matching_data_node["url"], session)
+            if status in [200, 302, 308]:
                 picked_data_node = data_node
                 print(f"DEBUG: Picking preferred data_node: {picked_data_node}")
                 break
             else:
                 print(f"Got status {status} for {matching_data_node['instance_id']}")
         elif len(matching_data_nodes) == 0:
-            print(f'DEBUG: Data node: {data_node} not available')
+            print(f"DEBUG: Data node: {data_node} not available")
```
[the remark this answers is not in the document]
Comment (jbusecke) [reply pangeo_forge_esgf/recipe_inputs.py:253]: #5 would be ready to merge. Let me know if it would help merging that before. 
I personally would have a much easier time to focus on the functional changes you made here if we can get rid of the formatting stuff.

[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/recipe_inputs.py:253]: Yes, absolutely! I woud very much appreciate the pre-commit config!

[the remark this answers is not in the document]
Comment (jbusecke) [reply pangeo_forge_esgf/recipe_inputs.py:253]: Ok let me merge #5 and then you can rebase this.

**Comment by larsbuntemeyer** [comment]
Sorry, i am not sure if this added unnecessary commits...

File: pangeo_forge_esgf/params.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/params.py:3]: I think we could also access the time frequency facet although it's not part of the dataset_id. It still has different names for the ESGF projects, (e.g. `time_frequency` in CORDEX but `frequency` in CMIP6). In general, the CORDEX vocabulary is derived from CMIP5 vocabulary. However, this could easily become a class attribute as you suggested...

File: pangeo_forge_esgf/recipe_inputs.py [hunk 12]
```diff
@@ -160,8 +164,15 @@ async def _esgf_api_request(
...
+
     if len(resp_data) == 0:
         raise ValueError(f"No Files were found for {iid}")
+    # Since we have hacked CORDEX special case above, i'll do it here again:
+    # rename to common CMIP vocabulary if neccessary
+    resp_data = [
+        {"frequency" if k == "time_frequency" else k: v for k, v in r.items()}
+        for r in resp_data
+    ]
     return resp_data
```
Comment (larsbuntemeyer) [inline pangeo_forge_esgf/recipe_inputs.py:176]: Sorry, this was neccessary since frequency facets have different names in CMIP6 and CMIP5 and CORDEX. I thinks this could really be tackled more easily in an object orient approach.

File: pangeo_forge_esgf/dynamic_kwargs.py [hunk 13]
```diff
@@ -96,21 +98,29 @@ async def response_data_processing(
...
     sizes = [r["size"] for r in response_data]
     titles = [r["title"] for r in response_data]
 
+    # determine frequency
+    frequencies = [r["frequency"][0] for r in response_data]
+    # check if not all files have the same frequency, that should never happen.
+    if not all(f == frequencies[0] for f in frequencies):
+        raise ValueError("Determined frequencies are not all equal.")
+    frequency = frequencies[0]
+
```
Comment (larsbuntemeyer) [inline pangeo_forge_esgf/dynamic_kwargs.py:115]: I added `frequency` (or `time_frequency`) to the filtered response. This should also work for `obs4MIPs` or `input4MIPs`...

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -0,0 +1,21 @@
+default_language_version:
```
Comment (jbusecke) [inline .pre-commit-config.yaml:1]: Is there a way to get rid of these commits? These have been introduced by merging an earlier commit and are not part of this PR, right?

File: pangeo_forge_esgf/params.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+cordex_params = {
+    "type": "Dataset",
+    "format": "application/solr+json",
+    "fields": "instance_id",
+    "latest": "true",
+    "distrib": "true",
+    "limit": 500,
+}
```
Comment (jbusecke) [inline pangeo_forge_esgf/params.py:23]: Lots of repeated code here. As far as I can see the only different is `"retracted": "false"`? Could you refactor this to a 'base_params` dict, that is then updated for each project?

**Comment by jbusecke** [comment]
Great progress on this. Thank you so much for the continued effort. I am a bit concerned that we need some very basic tests here (I actually have introduced a basic test setup in #8). I think we should have at least a set of tests for e.g. CMIP6/CMIP5/CORDEX to ensure that for a dummy dataset of the same dimensions we are getting the same answer out of a call to `response_data_processing`? Happy to help with this in the coming days if that would be useful. 

Alternatively I could work on another PR that implements such a high level test for CMIP6 only, you could merge that PR and then add tests for CORDEX here?

File: pangeo_forge_esgf/recipe_inputs.py (hunk 1 above)
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/recipe_inputs.py:150]: I don't really know.. I think in CORDEX, datasets are always the "latest" version while retracted versions are removed from ESGF...

File: README.md (hunk 8 above)
Comment (larsbuntemeyer) [reply README.md:51]: Yes, that was just to show, that you can choose a different url. CORDEX searches should actually also work for all ESGF nodes.

File: pangeo_forge_esgf/dynamic_kwargs.py (hunk 9 above)
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/dynamic_kwargs.py:101]: Agreed!

File: pangeo_forge_esgf/params.py (hunk 15 above)
Comment (larsbuntemeyer) [reply pangeo_forge_esgf/params.py:23]: Yes, absolutely, no problem!

**Comment by larsbuntemeyer** [comment]
I would really appreciate some tests, so I am very grateful for #8 . Concerning further tests, I think I could probably adapt CMIP6 tests for CORDEX if you like.

I have to admit, i am never really sure whether to rebase or merge the main branch after it has been updated. I know the difference, but would you like to keep all commits in them main? Or would you squash the PR? I would be grateful for some advice so i don't add unnecessary merge commits...

**Comment by jbusecke** [comment]
> I have to admit, i am never really sure whether to rebase or merge the main branch after it has been updated. I know the difference, but would you like to keep all commits in them main? 

Did you merge or rebase in the first place? I am also always confused about this situation TBH. Ill check a bit more how to resolve this as my go-to solution (make another PR) is not really satisfactory...

> Or would you squash the PR? I would be grateful for some advice so i don't add unnecessary merge commits...

I usually squash all commits in the PR, this was mostly for the sake of easier reviewing.

**Comment by jbusecke** [comment]
I just found [this](https://stackoverflow.com/a/51400593), but I am not sure if a force push will destroy our exchanges here.
Overall I think the tests are a priority here. I would suggest to add these first and then try to clean up the commit history for a final review? Thank you again for the work and sorry for the delay.

**Comment by larsbuntemeyer** [comment]
Yes, agreed! I'll have some time for this at the end of the week, they i will try to clean up a little bit.

**Comment by larsbuntemeyer** [comment]
Just some thoughts here: I have to rethink the `mip` keyword, it does not make sense. It should be `project` which is the API keyword. There is an additional complication in the CORDEX project since there is an additional `project` keyword in the facet search, e.g., there can be `project=cordex` where the `project` facet is hardcode in the `dataset_id_template`:
```json
"id":"cordex.output.EUR-44.MPI-CSC.MPI-M-MPI-ESM LR.historical.r1i1p1.REMO2009.v1.mon.tas.v20150609|esgf1.dkrz.de",
"dataset_id_template_":["cordex.%(product)s.%(domain)s.%(institute)s.%(driving_model)s.%(experiment)s.%(ensemble)s.%(rcm_name)s.%(rcm_version)s.%(time_frequency)s.%(variable)s"]
```
or `project=cordex-reklies` where the `project` facet can be parsed:
```json
"id":"cordex-reklies.output.EUR-11.GERICS.MIROC-MIROC5.historical.r1i1p1.REMO2015.v1.mon.tas.v20170329|esgf1.dkrz.de",
 "dataset_id_template_":["%(project)s.%(product)s.%(domain)s.%(institute)s.%(driving_model)s.%(experiment)s.%(ensemble)s.%(rcm_name)s.%(rcm_version)s.%(time_frequency)s.%(variable)s"]
```
I have to make sure to search all cordex project not only `project=cordex`...

**Comment by larsbuntemeyer** [comment]
More thoughts: 

If i chose `project=CORDEX-Reklies`, the `product` facet gets inactive, that really weired since it is still part of the dataset_id:

grafik

**Comment by jbusecke** [comment]
> If i chose project=CORDEX-Reklies, the product facet gets inactive:

Is there a value when you query the API directly? I wonder if this is some glitch in the GUI

Also is there a list of all available projects? I have to admit we are moving into somewhat unfamiliar territory for me here, and I hope that having a full list will give us an idea how to most efficiently generalize this.

**Comment by larsbuntemeyer** [comment]
I'll try to remove some of the `if` statements that are hacked in depending on the project. I think, this could then be more generalized. The ESGF CORDEX project is not always as consistent as it could be, that's why there might be come exceptions. Sorry for the late responses here, i had to dig through some old grumpy Fortran code the last week...

**Comment by larsbuntemeyer** [comment]
Although the recipe run works locally, feedstock creation is not yet possible due to https://github.com/pangeo-forge/pangeo-forge-recipes/issues/79 and https://github.com/pangeo-forge/pangeo-forge-recipes/issues/418.

**Comment by jbusecke** [comment]
Hey @larsbuntemeyer just checking in here (sorry for the long silence). I just wanted to make you aware of some new developments in PGF-recipes (see https://github.com/jbusecke/pangeo-forge-esgf/pull/9#issuecomment-1655640407) that in my mind now completely eliminate the need for the  `dynamic_kwargs.py` module. I am planning to fully deprecate this logic and move towards handling those issues in the recipe generation itself. 

I have started (and will continue) to work on #15 which reflects these plans. I would be curious if you agree that the upstream changes eliminate your need for the dynamic_kwargs module as well?

**Comment by jbusecke** [comment]
Hey @larsbuntemeyer just digging out of a mountain of gh notifications and found this. I have quite substantially refactored this package since the new pangeo-forge version (based on apache beam) became available. 

- I removed the `dynamic_kwargs.py` entirely. The process was very imprecise and slow anyways. We can actually now dynamically chunk by providing a custom function to the pangeo-forge-recipe stage. Check out this recent presentation for some details. I have made a new repo specifically for chunking algorithms, so if you had some custom logic in mind, please check it out: https://github.com/jbusecke/dynamic_chunks
- I also substantially refactored the ESGF client. This was all really done to serve my CMIP6 ingestion (), which now works really well, but I have to apologize for moving fast and destructive here. I think we could still accomodate CORDEX data here, but I think this would have to be a new PR? 

Please let me know if you are still interested in this, or have (understandibly) moved on.

**Comment by larsbuntemeyer** [comment]
Hey @jbusecke thanks for the updates! I have been following a bit your work and just watched your [Pangeo showcase](https://www.youtube.com/watch?v=vZKlcsYNNbU&list=PLuQQBBQFfpgq0OvjKbjcYgTDzDxTqtwua&index=3) recently, great work! I have been moving on a bit with [pange-forge-cordex](https://github.com/euro-cordex/pangeo-forge-cordex) to ingest [CORDEX data on AWS](https://registry.opendata.aws/euro-cordex/).

However, this is all based on pangeo-forge-recipes < 0.10.0, so instead of refactoring that, i will check out your work and see how far i get. But a new PR will definitely make sense, so this one should be closed!
